src/data_structures/fmindex.rs

```diff
@@ -88,41 +100,78 @@ pub trait FMIndexable {
 
 }
 
-/// The Fast Index in Minute space (FM-Index, Ferragina and Manzini, 2000) for finding suffix array
-/// intervals matching a given pattern.
+pub trait FMIndexCore {
+    type SA: SuffixArray;
 
-#[cfg_attr(feature = "serde_macros", derive(Serialize, Deserialize))]
-pub struct FMIndex<
-        DBWT: DerefBWT + Clone,
-        DLess: DerefLess + Clone,
-        DOcc: DerefOcc + Clone> {
-    bwt: DBWT,
-    less: DLess,
-    occ: DOcc,
+    fn occ(&self, r: usize, a: u8) -> usize;
+    fn less(&self, a: u8) -> usize;
+    fn bwt(&self) -> &BWT;
+    fn sa(&self) -> &Self::SA;
 }
 
-impl<
-    DBWT: DerefBWT + Clone,
-    DLess: DerefLess + Clone,
-    DOcc: DerefOcc + Clone> FMIndexable for FMIndex<DBWT, DLess, DOcc> {
+impl FMIndexCore for SampledSuffixArray {
+    // the trait needs to track what kind of suffix array reference it's returning
+    type SA = SampledSuffixArray;
```

Comment:
The FM-index allows users to perform a number of handy algorithms that result in suffix array intervals. It's true that to convert back to the original text positions, you'll need the suffix array. However, it's not uncommon for these steps to be done by separate actors which may or may not have access to the same memory. These steps may even be done sequentially by separate programs which read/write to a file.